### PR TITLE
Femdem/correcting elements to mapping procedure

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
In this PR I'm correcting details of the elements and main in order to perform better when remeshing

### CODE REVIEW

File: applications/FemToDemApplication/custom_elements/femdem2d_element.cpp [hunk 1]
```diff
@@ -792,31 +792,25 @@ void FemDem2DElement::CalculateOnIntegrationPoints(
...
-	}
-
-	// Computation of the l_char
-	if (aux < 3) {  // It is not an edge element --> The 2 elements are not equal
-		return std::pow((std::pow(Xcoord[0] - Xcoord[1], 2) + std::pow(Ycoord[0] - Ycoord[1], 2)), 0.5); // Length of the edge between 2 elements                                                                  // Currently the characteristic length is the edge length (can be modified)
-	} else { // Edge element
-		const double element_area = std::abs(this->GetGeometry().Area());
-		return std::sqrt(4.0 * element_area / std::sqrt(3.0)); // Cervera's Formula
-	} // l_char computed
+	Geometry<Node<3>>& nodes_current_element = pCurrentElement->GetGeometry(); // 3 nodes of the Element 1
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem2d_element.cpp]: r_nodes...

Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_elements/femdem2d_element.cpp]: jej Bulseye

Comment (loumalouomega) [reply applications/FemToDemApplication/custom_elements/femdem2d_element.cpp]: Try to define Geometry<Node<3>> as GeometryType, maybe some day the template argument of Node will disapear..

File: applications/FemToDemApplication/custom_elements/femdem2d_element.cpp [hunk 2]
```diff
@@ -1261,6 +1255,12 @@ void FemDem2DElement::SimoJuCriterion(
 	const double F = uniaxial_stress - rThreshold;
 
 	const double A = 1.00 / (Gt * n * n * E / (Length * std::pow(sigma_c, 2)) - 0.5);
+	if (A < tolerance) {
+		KRATOS_WATCH(this->Id())
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem2d_element.cpp]: Use the log please

File: applications/FemToDemApplication/custom_processes/compute_normalized_free_energy_on_nodes_process.hpp [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+        const double length1 = MathUtils<double>::Norm3(node_1_coordinates-node_2_coordinates);
+        const double length2 = MathUtils<double>::Norm3(node_2_coordinates-node_3_coordinates);
+        const double length3 = MathUtils<double>::Norm3(node_3_coordinates-node_1_coordinates);
+        return (length1 + length2 + length3) / 3.0;
+	}
+
+	double CalculateCharacteristicLength3D(const Geometry<Node<3>>& rGeometry)
+	{
+        Matrix node_indices;
+		node_indices.resize(6, 2);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/compute_normalized_free_energy_on_nodes_process.hpp]: What is this?, can you iterate over Edges of the geometry?

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Look my comments

File: applications/FemToDemApplication/custom_processes/compute_normalized_free_energy_on_nodes_process.hpp (hunk 3 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_processes/compute_normalized_free_energy_on_nodes_process.hpp]: yes... It's a loop over the edges. This node_indices are the local id of the nodes of each edge

Comment (loumalouomega) [reply applications/FemToDemApplication/custom_processes/compute_normalized_free_energy_on_nodes_process.hpp]: Please use the Edges from the geometry, what happen if we swap two edges in the geometry definition?

Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_processes/compute_normalized_free_energy_on_nodes_process.hpp]: Ok I'm going to change that

Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_processes/compute_normalized_free_energy_on_nodes_process.hpp]: Now I suppose that the code is performing the same but more consistently

**Comment by AlejandroCornejo** [comment]
> Look my comments

Done!

File: applications/FemToDemApplication/custom_processes/compute_normalized_free_energy_on_nodes_process.hpp [hunk 4]
```diff
@@ -165,12 +165,13 @@ class ComputeNormalizedFreeEnergyOnNodesProcess : public Process
 
     	Vector lengths = ZeroVector(6);
         for (unsigned int edge = 0; edge < 6; edge++) { // Loop over edges
-            const double X1 = rGeometry.Edges()[edge][0].X0();
-            const double X2 = rGeometry.Edges()[edge][1].X0();
-            const double Y1 = rGeometry.Edges()[edge][0].Y0();
-            const double Y2 = rGeometry.Edges()[edge][1].Y0();
-            const double Z1 = rGeometry.Edges()[edge][0].Z0();
-            const double Z2 = rGeometry.Edges()[edge][1].Z0();
+            auto& edges = rGeometry.Edges();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/compute_normalized_free_energy_on_nodes_process.hpp]: r_edge, and declare it out of the loop!!!

**Review by loumalouomega — CHANGES_REQUESTED** [review]

File: applications/FemToDemApplication/custom_processes/compute_normalized_free_energy_on_nodes_process.hpp (hunk 4 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_processes/compute_normalized_free_energy_on_nodes_process.hpp]: yeah, sorry

**Review by loumalouomega — APPROVED** [review]

**Review by loumalouomega — APPROVED** [review]

**Pull request merged**